src/Microsoft.Restier.Core/ApiConfiguration.cs

```diff
@@ -65,5 +68,46 @@ public IServiceProvider ServiceProvider
         {
             return this.serviceProvider.GetService<T>();
         }
+
+        internal TaskCompletionSource<IEdmModel> CompeteModelGeneration(out Task<IEdmModel> running)
+        {
+            var source = new TaskCompletionSource<IEdmModel>(TaskCreationOptions.AttachedToParent);
+            var exist = Interlocked.CompareExchange(ref modelTask, source.Task, null);
+            if (exist != null)
+            {
+                running = exist;
+                source.SetCanceled();
+                return null;
+            }
+
+            source.Task.ContinueWith(
```

Comment:
Sorry I mean `Task.ContinueWith` returns a continuation task which will be called when, as you said, `SetResult` or `SetException` is called. But does it have a side effect that the original task is appended with the continued task as well? If not, `source.Task` will return the original one rather than the continued one. But that may not be a problem though.